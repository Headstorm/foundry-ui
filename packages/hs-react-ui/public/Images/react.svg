<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 19.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_2" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 width="561.8px" height="499.9px" viewBox="-80.4 151.2 561.8 499.9" style="enable-background:new -80.4 151.2 561.8 499.9;"
	 xml:space="preserve">
<style type="text/css">
	.st0{fill:#939598;}
</style>
<circle class="st0" cx="200.5" cy="400.6" r="50.2"/>
<path class="st0" d="M200.5,515.6c-70.5,0-132.1-8.3-178.2-24.1c-29.9-10.2-55.3-23.8-73.4-39.3c-19.2-16.4-29.4-34.3-29.4-51.6
	c0-33.2,36.4-65.7,97.5-86.9c50-17.4,115.2-27.1,183.4-27.1c67,0,131.3,9.4,181,26.6c29.1,10,53.6,23,71,37.4
	c18.9,15.8,28.9,33.1,28.9,50c0,34.5-40.7,69.4-106.3,91.1C328.6,507.1,266.6,515.6,200.5,515.6z M200.5,310.6
	c-64.7,0-128.7,9.4-175.5,25.7c-40.5,14.1-81.4,39.5-81.4,64.3c0,10,7.4,21.8,21,33.4c15.8,13.5,38.4,25.6,65.5,34.8
	c43.6,14.9,102.6,22.8,170.4,22.8c63.6,0,122.9-8,167-22.7c61.7-20.5,89.9-49.8,89.9-68.3c0-9.5-7.2-20.7-20.3-31.6
	c-15.1-12.6-37.1-24.1-63.4-33.2C326.4,319.6,264.9,310.6,200.5,310.6z"/>
<path class="st0" d="M86.6,650.8c-10.2,0-19.2-2.2-26.8-6.6C31,627.6,21.1,579.8,33.1,516.3c9.9-52.1,34.1-113.3,68.2-172.4h0
	c33.5-58,73.7-109,113.4-143.5c23.2-20.2,46.7-35,67.9-42.8c23.1-8.5,43.1-8.5,57.7-0.1c29.9,17.2,39.8,70,25.8,137.6
	c-9.9,48-33.5,105.9-66.5,163.2c-35.2,61-73.2,110.2-109.9,142.3c-23.8,20.8-48.3,36-70.7,43.9C107.4,648.7,96.4,650.8,86.6,650.8z
	 M122.1,355.9c-32.3,56-56.2,116.1-65.4,164.9c-8,42.1-6.5,90.2,15,102.6c16.1,9.3,55,0.5,102.3-40.9c34.7-30.3,71-77.5,104.9-136.2
	c31.8-55.1,54.4-110.5,63.8-156c6-29.1,7.3-55.5,3.8-76.4c-3-17.9-9.4-30.6-18.2-35.6c-8.2-4.7-21.5-4.1-37.5,1.8
	c-18.5,6.8-39.4,20.1-60.4,38.4C192.8,251.3,154.3,300.1,122.1,355.9L122.1,355.9z"/>
<path class="st0" d="M314.4,651.1c-27.2,0-61.7-16.4-97.7-47.4c-40.2-34.6-81.1-86.1-115.3-145.2v0c-33.6-58-57.6-118.3-67.7-170
	c-5.9-30.2-7-57.9-3.2-80.2c4.2-24.3,14.1-41.6,28.8-50.1c29.8-17.3,80.5,0.5,132.1,46.4c36.6,32.5,75,81.9,108.1,139.1
	c35.3,61,59,118.5,68.4,166.3c6.1,31,7.1,59.8,2.8,83.2c-4.6,24.9-15,42.6-30,51.3C333.2,648.9,324.3,651.1,314.4,651.1z
	 M122.2,446.5c32.4,56,72.6,106.7,110.2,139c32.5,28,74.9,50.6,96.4,38.2c16.1-9.3,27.9-47.4,15.7-109
	c-9-45.2-31.7-100.2-65.6-158.9c-31.9-55.1-68.6-102.4-103.3-133.2c-22.2-19.7-44.4-34.1-64.2-41.5c-17-6.4-31.2-7.1-39.9-2.1
	c-8.2,4.7-14.3,16.6-17.2,33.4c-3.3,19.4-2.3,44.2,3.1,71.5C66.9,333,89.9,390.7,122.2,446.5L122.2,446.5z"/>
</svg>
