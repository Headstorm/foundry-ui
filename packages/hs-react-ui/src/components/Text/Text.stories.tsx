import React from 'react';
import styled from 'styled-components';
import { color as color2, text, boolean, select } from '@storybook/addon-knobs';

import Icon from '@mdi/react';
import * as Icons from '@mdi/js';
import Text from './Text';
import colors from '../../enums/colors';
import timings from '../../enums/timings';

export default {
  title: 'Text',
  component: Text,
};

const themeColors = {
  ...colors,
  background: 'beige',
  primary: 'purple',
};

const themeTimings = {
  ...timings,
  xSlow: '2s',
};

const options = {
  none: '',
  ...Icons,
};

const getIconPath = (path: string) => (path ? <Icon size="16px" path={path} /> : undefined);

const ThemedContainer = styled.div`
  ${({ size, color }: { size: number | string; color: string }) => `
    width: fit-content;
    background-color: ${themeColors.background};
    transition: transform ${themeTimings.xSlow};
    font-family: Roboto, sans-serif;
    font-size: ${size};
    color: ${color};
    border: 1px solid ${themeColors.primary};
  `}
`;

/* Default */

export const Default = () => (
  <Text
    size={text('size', '1rem')}
    color={color2('color', colors.grayXlight)}
    iconPrefix={getIconPath(select('iconPrefix', options, options.mdiComment))}
    iconSuffix={getIconPath(select('iconSuffix', options, options.mdiComment))}
    isProcessing={boolean('isProcessing', false)}
    isLoading={boolean('isLoading', false)}
  >
    {text('children', 'Lorem ipsum dolor sit amet.')}
  </Text>
);

/* Themed with a static Icon */

export const ThemedTextWithStaticIcons = () => (
  <Text
    StyledContainer={ThemedContainer}
    size={text('size', '1rem')}
    color={color2('color', colors.grayDark)}
    iconPrefix={getIconPath(select('iconPrefix', options, options.mdiComment))}
    iconSuffix={getIconPath(select('iconPrefix', options, options.mdiComment))}
    isProcessing={boolean('isProcessing', false)}
    isLoading={boolean('isLoading', false)}
  >
    {text('children', 'Lorem ipsum dolor sit amet.')}
  </Text>
);

/* Themed with all knobs */

export const ThemedText = () => (
  <Text
    StyledContainer={ThemedContainer}
    size={text('size', '1rem')}
    color={color2('color', colors.grayDark)}
    iconPrefix={getIconPath(select('iconPrefix', options, options.mdiComment))}
    iconSuffix={getIconPath(select('iconSuffix', options, options.mdiComment))}
    isProcessing={boolean('isProcessing', false)}
    isLoading={boolean('isLoading', false)}
  >
    {text('children', 'Lorem ipsum dolor sit amet.')}
  </Text>
);
