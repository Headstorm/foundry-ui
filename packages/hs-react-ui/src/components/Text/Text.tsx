import React from 'react';
import styled, { StyledComponentBase } from 'styled-components';

import Icon from '@mdi/react';
import { mdiLoading } from '@mdi/js';
import Progress from '../Progress/Progress';
import { Span } from '../../htmlElements';
import { SubcomponentPropsType } from '../commonTypes';

/* Default Styled Text Container */
export const TextContainer = styled(Span)`
  ${({ size, color }: { size: string; color: string }) => `
    font-size: ${size};
    color: ${color};
  `}
`;

export interface TextProps {
  children?: string | Node;
  color?: string;
  iconPrefix?: string | JSX.Element;
  iconSuffix?: string | JSX.Element;
  isLoading?: boolean;
  isProcessing?: boolean;
  size?: string;

  StyledContainer?: string & StyledComponentBase<any, {}>;
  StyledIconContainer?: string & StyledComponentBase<any, {}>;
  containerProps?: SubcomponentPropsType;
  iconContainerProps?: SubcomponentPropsType;
}

/* Styled div that represents the scroll bar */
const StyledProgress = styled(Progress)`
  ${({ size }: { size: string }) => `
    width: calc(${size} * 10);
    height: ${size};
  `}
`;

const IconContainer = styled(Span)`
  ${({ side }: { side: 'left' | 'right' }) => `
    margin-${side === 'left' ? 'right' : 'left'}: .5em;
    display: inline-flex;
    vertical-align: middle;
  `}
`;

// TODO: If children are passed in and loading===true,
// set the children to visibility: none and have the loading bar fill their width
// that way the end developer can pass "placeholder" text for how long the text will probably be
// when loaded
const Text = ({
  children,
  color,
  iconPrefix,
  iconSuffix,
  isLoading,
  isProcessing,
  size = '1rem',

  StyledContainer = TextContainer,
  StyledIconContainer = IconContainer,
  containerProps = {},
  iconContainerProps = {},
}: TextProps) => (
  <StyledContainer data-test-id="hsui-Text" size={size} color={color} {...containerProps}>
    {isLoading && <StyledProgress size={size} />}
    {!isLoading &&
      !isProcessing &&
      iconPrefix &&
      (typeof iconPrefix === 'string' && iconPrefix !== '' ? (
        <StyledIconContainer side="left" {...iconContainerProps}>
          <Icon path={iconPrefix} size={size} />
        </StyledIconContainer>
      ) : (
        <StyledIconContainer side="left" size={size} {...iconContainerProps}>
          {iconPrefix}
        </StyledIconContainer>
      ))}
    {!isLoading && isProcessing && (
      <StyledIconContainer side="left" {...iconContainerProps}>
        <Icon path={mdiLoading} size={size} spin={1} />
      </StyledIconContainer>
    )}
    {!isLoading && children}

    {!isLoading &&
      iconSuffix &&
      (typeof iconSuffix === 'string' ? (
        <StyledIconContainer side="right" {...iconContainerProps}>
          <Icon path={iconSuffix} size={size} />
        </StyledIconContainer>
      ) : (
        <StyledIconContainer side="right" {...iconContainerProps}>
          {iconSuffix}
        </StyledIconContainer>
      ))}
  </StyledContainer>
);

Text.Container = TextContainer;
Text.IconContainer = IconContainer;
export default Text;
