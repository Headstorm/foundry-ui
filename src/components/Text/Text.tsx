import React, { ReactNode, RefObject } from 'react';
import styled from 'styled-components';

import Icon from '@mdi/react';
import { mdiLoading } from '@mdi/js';
import Skeleton from '../Skeleton/Skeleton';
import { Span } from '../../htmlElements';
import { SubcomponentPropsType, StyledSubcomponentType } from '../commonTypes';

/* Default Styled Text Container */
export const TextContainer = styled(Span)`
  ${({ size, color }: { size: string; color: string }) => `
    font-size: ${size};
    color: ${color};
  `}
`;

export interface TextProps {
  children?: ReactNode;
  color?: string;
  iconPrefix?: string | JSX.Element;
  iconSuffix?: string | JSX.Element;
  isLoading?: boolean;
  isProcessing?: boolean;
  size?: string;

  StyledContainer?: StyledSubcomponentType;
  StyledIconContainer?: StyledSubcomponentType;
  containerProps?: SubcomponentPropsType;
  iconContainerProps?: SubcomponentPropsType;
  containerRef?: RefObject<HTMLDivElement>;
  iconPrefixContainerRef?: RefObject<HTMLElement>;
  iconSuffixContainerRef?: RefObject<HTMLElement>;
}

const SkeletonContainer = styled(Skeleton.Container)`
  display: inline;
`;

const IconContainer = styled(Span)`
  ${({ side }: { side: 'left' | 'right' }) => `
    margin-${side === 'left' ? 'right' : 'left'}: .5em;
    display: inline-flex;
    vertical-align: middle;
  `}
`;

// TODO: If children are passed in and loading===true,
// set the children to visibility: none and have the loading bar fill their width
// that way the end developer can pass "placeholder" text for how long the text will probably be
// when loaded
const Text = ({
  children,
  color,
  iconPrefix,
  iconSuffix,
  isLoading,
  isProcessing,
  size = '1rem',

  StyledContainer = TextContainer,
  StyledIconContainer = IconContainer,
  containerProps = {},
  iconContainerProps = {},
  containerRef,
  iconPrefixContainerRef,
  iconSuffixContainerRef,
}: TextProps): JSX.Element => (
  <StyledContainer
    data-test-id="hsui-Text"
    size={size}
    color={color}
    ref={containerRef}
    {...containerProps}
  >
    {!isProcessing &&
      iconPrefix &&
      (typeof iconPrefix === 'string' && iconPrefix !== '' ? (
        <StyledIconContainer side="left" ref={iconPrefixContainerRef} {...iconContainerProps}>
          <Icon path={iconPrefix} size={size} />
        </StyledIconContainer>
      ) : (
        <StyledIconContainer
          side="left"
          size={size}
          ref={iconPrefixContainerRef}
          {...iconContainerProps}
        >
          {iconPrefix}
        </StyledIconContainer>
      ))}
    {isProcessing && (
      <StyledIconContainer side="left" ref={iconPrefixContainerRef} {...iconContainerProps}>
        <Icon aria-hidden="true" path={mdiLoading} size={size} spin={1} />
      </StyledIconContainer>
    )}
    <Skeleton color={color} StyledContainer={SkeletonContainer} isLoading={isLoading}>
      <Span>{children}</Span>
    </Skeleton>

    {iconSuffix &&
      (typeof iconSuffix === 'string' ? (
        <StyledIconContainer side="right" ref={iconSuffixContainerRef} {...iconContainerProps}>
          <Icon path={iconSuffix} size={size} />
        </StyledIconContainer>
      ) : (
        <StyledIconContainer side="right" ref={iconSuffixContainerRef} {...iconContainerProps}>
          {iconSuffix}
        </StyledIconContainer>
      ))}
  </StyledContainer>
);

Text.Container = TextContainer;
Text.IconContainer = IconContainer;
export default Text;
