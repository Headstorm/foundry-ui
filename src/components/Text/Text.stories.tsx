import React from 'react';
import { Story, Meta } from '@storybook/react';

import Icon from '@mdi/react';
import * as Icons from '@mdi/js';

import Text from './Text';
import colors from '../../enums/colors';

const options = {
  none: '',
  ...Icons,
};

export const Default: Story = args => {
  const getIconPath = (path: string) => (path ? <Icon size={args.size} path={path} /> : undefined);

  return (
    <Text
      {...args}
      iconPrefix={getIconPath(args.iconPrefix)}
      iconSuffix={getIconPath(args.iconSuffix)}
    ></Text>
  );
};
Default.args = {
  size: '1rem',
  color: colors.grayDark,
  iconPrefix: 'mdiComment',
  iconSuffix: 'mdiComment',
  isProcessing: false,
  isLoading: false,
  children: 'Lorem ipsum dolor sit amet.',
};

const iconOptions = {
  options: Object.keys(options),
  mapping: options,
  control: {
    type: 'select',
  },
};

export default {
  title: 'Text',
  component: Text,
  argTypes: {
    iconPrefix: iconOptions,
    iconSuffix: iconOptions,
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/3r2G00brulOwr9j7F6JF59/Generic-UI-Style?node-id=83%3A17',
    },
  },
} as Meta;
