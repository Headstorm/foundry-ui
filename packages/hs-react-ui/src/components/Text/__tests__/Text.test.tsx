import React from 'react';
import { render, fireEvent, waitFor, configure } from '@testing-library/react';
import Icon from '@mdi/react';
import Text from '../Text';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);
configure({ testIdAttribute: 'data-test-id' });

const testId = 'hsui-Text';

describe('Text', () => {
  it('matches snapshot', async () => {
    const { container, getByTestId } = render(<Text />);
    await waitFor(() => getByTestId(testId));
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with isLoading', async () => {
    const { container, getByTestId } = render(<Text isLoading />);
    await waitFor(() => getByTestId(testId));
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with isProcessing', async () => {
    const { container, getByTestId } = render(<Text isProcessing />);
    await waitFor(() => getByTestId(testId));
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with icon strings', async () => {
    const { container, getByTestId } = render(
      <Text iconPrefix="mdiComment" iconSuffix="mdiComment" />,
    );
    await waitFor(() => getByTestId(testId));
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with icons', async () => {
    const { container, getByTestId } = render(
      <Text iconPrefix={<Icon path={'mdiComment'} />} iconSuffix={<Icon path={'mdiComment'} />} />,
    );
    await waitFor(() => getByTestId(testId));
    expect(container).toMatchSnapshot();
  });
  describe('Accessibility Tests', () => {
    it('Should pass accessibility test with default props', async () => {
      const component = <Text></Text>;
      const { container } = render(component);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
  describe('Ref tests', () => {
    it('containerRef.current should exist', async () => {
      const testId = 'hsui-Text';
      const containerProps = { 'data-test-id': testId };
      const ref = React.createRef<HTMLSpanElement>();
      const { getByTestId } = render(
        <Text containerRef={ref} containerProps={containerProps}>
          test
        </Text>,
      );
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLSpanElement).toBeTruthy();
    });
    it('iconPrefixContainerRef.current should exist', async () => {
      const testId = 'hsui-Text';
      const containerProps = { 'data-test-id': testId };
      const ref = React.createRef<HTMLElement>();
      const { getByTestId } = render(
        <Text
          iconPrefixContainerRef={ref}
          iconPrefix={<div>prefix</div>}
          containerProps={containerProps}
        />,
      );
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLElement).toBeTruthy();
    });
    it('iconSuffixContainerRef.current should exist', async () => {
      const testId = 'hsui-Text';
      const containerProps = { 'data-test-id': testId };
      const ref = React.createRef<HTMLElement>();
      const { getByTestId } = render(
        <Text
          iconSuffixContainerRef={ref}
          iconSuffix={<div>prefix</div>}
          containerProps={containerProps}
        />,
      );
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLElement).toBeTruthy();
    });
  });
});
