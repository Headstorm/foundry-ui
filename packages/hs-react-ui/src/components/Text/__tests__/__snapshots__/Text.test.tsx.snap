// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text matches snapshot 1`] = `
.c0 {
  font-size: 1rem;
  color: undefined;
}

<div>
  <span
    class="c0"
    data-test-id="hsui-Text"
  />
</div>
`;

exports[`Text matches snapshot with icon strings 1`] = `
.c0 {
  font-size: 1rem;
  color: undefined;
}

.c1 {
  margin-right: .5em;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-bottom;
}

.c2 {
  margin-left: .5em;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-bottom;
}

<div>
  <span
    class="c0"
    data-test-id="hsui-Text"
  >
    <span
      class="c1"
    >
      <svg
        role="presentation"
        style="width: 1rem; height: 1rem;"
        viewBox="0 0 24 24"
      >
        <path
          d="mdiComment"
          style="fill: currentColor;"
        />
      </svg>
    </span>
    <span
      class="c2"
    >
      <svg
        role="presentation"
        style="width: 1rem; height: 1rem;"
        viewBox="0 0 24 24"
      >
        <path
          d="mdiComment"
          style="fill: currentColor;"
        />
      </svg>
    </span>
  </span>
</div>
`;

exports[`Text matches snapshot with icons 1`] = `
.c0 {
  font-size: 1rem;
  color: undefined;
}

.c1 {
  margin-right: .5em;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-bottom;
}

.c2 {
  margin-left: .5em;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-bottom;
}

<div>
  <span
    class="c0"
    data-test-id="hsui-Text"
  >
    <span
      class="c1"
    >
      <svg
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="mdiComment"
          style="fill: currentColor;"
        />
      </svg>
    </span>
    <span
      class="c2"
    >
      <svg
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="mdiComment"
          style="fill: currentColor;"
        />
      </svg>
    </span>
  </span>
</div>
`;

exports[`Text matches snapshot with isLoading 1`] = `
.c0 {
  font-size: 1rem;
  color: undefined;
}

.c1 {
  background: linear-gradient( 45deg, rgba(255,255,255,0.75), rgba(0,0,0,0.75), rgba(255,255,255,0.75), rgba(0,0,0,0.75), rgba(255,255,255,0.75) ) repeat;
  background-size: 400% 100%;
  width: 6rem;
  height: 16px;
  border-radius: 9999px;
  -webkit-animation: ia-dwnO 8s linear infinite;
  animation: ia-dwnO 8s linear infinite;
  line-height: 0;
  width: calc(1rem * 10);
  height: 1rem;
}

<div>
  <span
    class="c0"
    data-test-id="hsui-Text"
  >
    <div
      class="c1"
    />
  </span>
</div>
`;

exports[`Text matches snapshot with isProcessing 1`] = `
.c0 {
  font-size: 1rem;
  color: undefined;
}

.c1 {
  margin-right: .5em;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: text-bottom;
}

<div>
  <span
    class="c0"
    data-test-id="hsui-Text"
  >
    <span
      class="c1"
    >
      <svg
        role="presentation"
        style="width: 1rem; height: 1rem;"
        viewBox="0 0 24 24"
      >
        <style>
          @keyframes spin { to { transform: rotate(360deg) } }
        </style>
        <g
          style="animation: spin linear 1s infinite; transform-origin: center;"
        >
          <path
            d="M12,4V2A10,10 0 0,0 2,12H4A8,8 0 0,1 12,4Z"
            style="fill: currentColor;"
          />
          <rect
            fill="transparent"
            height="24"
            width="24"
          />
        </g>
      </svg>
    </span>
  </span>
</div>
`;
